SimPEG/regularization/coupling.py

```diff
@@ -0,0 +1,619 @@
+        \nabla \mathbf{m_1}_i \times \nabla \mathbf{m_2}_i \|^2
+
+    All methods assume that we are working with two models only.
+
+    '''
+    def __init__(self, mesh, indActive, mapping, **kwargs):
+
+        self.as_super.__init__(mesh, indActive, mapping, **kwargs)
+        self.map1, self.map2 = mapping.maps # Assume a map has been passed for each model.
+
+        regmesh = self.regmesh
+        self._Dx = regmesh.aveFx2CC @ regmesh.cellDiffx
+        self._Dy = regmesh.aveFy2CC @ regmesh.cellDiffy
+        if regmesh.mesh.dim == 3:
+            self._Dz = regmesh.aveFz2CC @ regmesh.cellDiffz
+
+        self._dim = regmesh.mesh.dim
+        assert self._dim in (2,3), 'Cross-Gradient is only defined for 2D or 3D'
+
+
+    def models(self, ind_models):
+        '''
+        Method to pass models to CrossGradient object and ensures models are
+        compatible for further use. Checks that models are of same size.
+
+        :param container of numpy.ndarray ind_models: [model1, model2,...]
+
+        rtype: list of numpy.ndarray models: [model1, model2,...]
+        return: models
+
+        '''
+        models = []
+        n = len(ind_models)
+        for i in range(n):
+            # check that the models are either a list, tuple, or np.ndarray
+            assert isinstance(ind_models[i], (list,tuple,np.ndarray))
+            if isinstance(ind_models[i], (list,tuple)):
+                ind_models[i] = np.array(ind_models[i])
+
+        # check if models are of same size
+        it = iter(ind_models)
+        length = len(next(it))
+        if not all(len(l)==length for l in it):
+            raise ValueError('not all models are of the same size!')
+
+        for i in range(n):
+            models.append(ind_models[i])
+
+        return models
+
+
+    def calculate_gradient(self, model, normalize=False):
+        '''
+        Calculate the spatial gradients of the model using central difference.
+
+        Concatenates gradient components into a single array.
+        [[x_grad1, y_grad1, z_grad1],
+         [x_grad2, y_grad2, z_grad2],
+         [x_grad3, y_grad3, z_grad3],...]
+
+        :param numpy.ndarray model: model
+
+        :rtype: numpy.ndarray
+        :return: gradient_vector: array where each row represents a model cell,
+                 and each column represents a component of the gradient.
+
+        '''
+
+        if self._dim == 2:
+
+            x_grad = self._Dx.dot(model)
+            y_grad = self._Dy.dot(model)
+            grad_list = np.column_stack((x_grad, y_grad))
+
+        elif self._dim == 3:
+
+            x_grad = self._Dx.dot(model)
+            y_grad = self._Dy.dot(model)
+            z_grad = self._Dz.dot(model)
+            grad_list = np.column_stack((x_grad, y_grad, z_grad))
+
+        if normalize:
+            norms = np.linalg.norm(grad_list, axis=-1)
+            norms[norms<self.grad_tol] = 1.0 # avoid exteremly large values of 1/norms
+            grad_list = grad_list/norms[:, None]
```

Comment:
I'm not sure this will return the normalized gradients. If we divide by 1.0, the gradients won't be normalized. The idea was that if the norms were extremely small, we could ignore them and set the gradients to 0.
I suggest adding a line `grad_list[norms<self.grad_tol] = 0.0`

Reply:
done